### PR TITLE
HC-376: publish datasets

### PR DESCRIPTION
**Pull request by DustinKLo**
related tickets: 
- https://hysds-core.atlassian.net/browse/HC-376
- https://jira.jpl.nasa.gov/browse/NSDS-2114

related PR(s): 
- https://github.com/hysds/grq2/pull/51
- https://github.com/hysds/hysds/pull/140
- https://github.jpl.nasa.gov/IEMS-SDS/nisar-pcm/pull/507
- https://github.jpl.nasa.gov/IEMS-SDS/CNM_product_delivery/pull/50

Jenkins test(s):
- real PGE: https://nisar-pcm-ci.jpl.nasa.gov/job/force-branches/job/mkarim-force_branch-E2E-test/286/parameters/
- reprocessing - https://nisar-pcm-ci.jpl.nasa.gov/job/force-branches/job/dustinlo-force_branch-E2E-test/101/
- forward - https://nisar-pcm-ci.jpl.nasa.gov/job/force-branches/job/dustinlo-force_branch-E2E-test/103/
- dev-e2e reprocessing (sanity test) - https://nisar-pcm-ci.jpl.nasa.gov/job/force-branches/job/dustinlo-force_branch-E2E-test/109/

changes:
- moved dataset ingest functions in `utils.py` to `dataset_ingest.py`
- created new `dataset_ingest_bulk.py` function to handle "transactional" dataet publishing
  - publishes all datasets to s3 **first** before indexing to ES
  - will handle rollback of multiple datasets if one happens to fail
  - will use the new GRQ2 dataset publishing endpoint for "bulk" indexing
- renamed duplicate function naming in `utils.py` and `dataset_ingest.py`
- `post_processors` in `job_worker.py` is now using the new transactional dataset publishing function
- moved `triage` function in `utils.py` to its own module to avoid a cyclical dependency
- fixed unit tests to reflect changes and removed unused mocks
- added new `GRQ_UPDATE_URL_BULK` to `celeryconfig.py`

cases handled that were raised by Hook:
- nominal case handling: `try-catch` on s3 upload fail and rollback ✓
  - new `publish_datasets` function will rollback s3
- terminated case during S3 upload ✓
  - datasets will be orphaned (w/o ES doc), on retry will delete previous s3 objects and start anew
- terminated case during ES ingest ✓
  - if ES's bulk API accepts all docs, then it will officially be published (in s3 and ES)
     - the job might not be marked as `job-completed` and will not queue the `user_rules`, but the datasets will be there regardless
  - if ES's bulk API rejects it, then the docs will be rolled back and will be the same as the above case
- retry via rabbitmq `redeliver=true` ✓
  - similar to the 2nd case
- retry via `job-level` retry ✓
  - similar to the 2nd case
- orphan finder (best case scenario should not be needed with these new changes)

### CODE REVIEW

**Comment by DustinKLo** [comment]
will wait until @pymonger gets back to do a final review before merging

File: test/test_triage.py
Comment (DustinKLo) [inline test/test_triage.py:84]: @mcayanan i had to move the `hysds.dataset_ingest.publish_dataset` mock to a decorator to get it to work, it was raising some error with file not found

are you okay with this solution? any ideas on how i can get it to work without using this decorator

Comment (mcayanan) [reply test/test_triage.py:84]: Per our tag up, let's see what @pymonger says. I'm not an expert on the mock stuff and I'm not sure why moving it out as a decorator worked. It was weird though that when the unit test was left as-is (the original implementation), that it looked like it was actually calling the `publish_dataset` function. It should have "mocked" it instead. My guess is that moving stuff around clobbered something to where basically the test doesn't know that this is function should be mocked.

Comment (pymonger) [reply test/test_triage.py:84]: Sorry just catching up on things. What was the error @DustinKLo?

Comment (DustinKLo) [reply test/test_triage.py:84]: it wasnt mocking one of the functions properly and says `electric` not found

File: hysds/dataset_ingest_bulk.py
Comment (pymonger) [inline hysds/dataset_ingest_bulk.py:71]: @DustinKLo: would we be able to use `get_func` from `hysds.utils`?

File: test/test_triage.py
Comment (pymonger) [reply test/test_triage.py:84]: @DustinKLo: the original test function shows:
```
        # Mocked data
        open_mock = umock.patch("hysds.utils.open", umock.mock_open()).start()
        makedirs_mock = umock.patch("os.makedirs").start()
        shutil_copy_mock = umock.patch("shutil.copy").start()
        publish_dataset_mock = umock.patch("hysds.utils.publish_dataset").start()
        publish_dataset_mock.return_value = {}
```
Perhaps try adding back the mock for `os.makedirs` and `shutil.copy` and removing the decorator?
```
       # Mocked data
        open_mock = umock.patch("hysds.triage.open", umock.mock_open()).start()
       makedirs_mock = umock.patch("os.makedirs").start()
        shutil_copy_mock = umock.patch("shutil.copy").start()
        publish_dataset_mock = umock.patch("hysds.dataset_ingest.publish_dataset").start()
        publish_dataset_mock.return_value = {}
```

File: hysds/dataset_ingest_bulk.py
Comment (DustinKLo) [reply hysds/dataset_ingest_bulk.py:71]: good catch, `get_func` still exists in `hysds.utils` so ill remove it here

**Comment by DustinKLo** [comment]
@pymonger @hookhua 
testing out PGEs to force a clobber issue and then re-trying with `_force_ingest` set to `True`
https://github.com/DustinKLo/test_pges/blob/master/area_of_interest/create_aoi.py

when re-running it will raise a `NoClobberException`

after adding `_force_ingest` to `_context.json` it will successfully publish the dataset and not raise a `NoClobberException`
https://hysds-core.atlassian.net/wiki/spaces/HYS/pages/199885846/Preprocessor+and+Postprocessor+Functions#publish_datasets
```python
with open(ctx_file) as f:
  ctx = json.load(f)
  
  ctx['_force_ingest'] = True
  
  with open(ctx_file, 'w') as f:
      json.dump(ctx, f, sort_keys=True, indent=2)
```

`_context.json`
```js
{
  "_command": "/home/ops/verdi/ops/test_pges/area_of_interest/create_aoi.sh",
  "_disk_usage": "100KB",
  "_force_ingest": true,  // <------- _force_ingest set here
  "_prov": {
    "wasGeneratedBy": "task_id:e6c2d441-6715-4038-b5d2-c97554e18b1e"
  },
  "container_image_id": "sha256:23118d490a3c8f337d172c06ec8c0221e2643d0275c4d8b4402bafb9687cfee4",
  "container_image_name": "container-dustinklo_test_pges:force_ingest",
  "container_image_url": "s3://s3-us-west-2.amazonaws.com/nisar-dev-cc-fwd-dustinlo/container-dustinklo_test_pges:force_ingest.tar.gz",
  "container_mappings": {},
  "container_specification": {
    "digest": "sha256:23118d490a3c8f337d172c06ec8c0221e2643d0275c4d8b4402bafb9687cfee4",
    "id": "container-dustinklo_test_pges:force_ingest",
    "url": "s3://s3-us-west-2.amazonaws.com/nisar-dev-cc-fwd-dustinlo/container-dustinklo_test_pges:force_ingest.tar.gz",
    "version": "force_ingest"
  },
  "job_priority": 0,
  "job_specification": {
    "command": "/home/ops/verdi/ops/test_pges/area_of_interest/create_aoi.sh",
    "container": "container-dustinklo_test_pges:force_ingest",
    "disk_usage": "100KB",
    "id": "job-aoi_oregon:force_ingest",
    "job-version": "force_ingest",
    "params": [],
    "recommended-queues": [
      "factotum-job_worker-small",
      "nisar-job_worker-small"
    ],
    "resource": "jobspec",
    "soft_time_limit": 900,
    "time_limit": 960
  },
  "localize_urls": [],
  "runtime_options": {},
  "tag": "force_ingest_true"
}
```

Screen Shot 2022-08-08 at 1 13 32 PM

**Review by mcayanan — APPROVED** [review]
LGTM

**Pull request merged**
